dot/rpc/websocketListeners.go

```diff
@@ -0,0 +1,371 @@
+
+	initRes := newSubscriptionResponseJSON(scl.subID, reqID)
+	c.safeSend(initRes)
+
+	//for k := range sub.Filter {
+	//	fmt.Printf("Lookup Key %v\n", k)
+	//
+	//	res, err := c.storageAPI.GetStorage(nil, common.MustHexToBytes(k))
+	//	if err != nil {
+	//		logger.Debug("error retrieving from storage", "error", err)
+	//	}
+	//	sub.Filter[k] = res
+	//}
+	// respond with current value
+	//res, err := c.storageAPI.GetStorage(nil, sub.)
+
+	return scl.subID, nil
+}
+
+// StorageChangeListener for listening to state change channels
+type StorageChangeListener struct {
+	channel chan *state.SubscriptionResult
+	wsconn  *WSConn
+	chanID  byte
+	subID   int
+}
+
+// Listen implementation of Listen interface to listen for importedChan changes
+func (l *StorageChangeListener) Listen() {
+	for change := range l.channel {
+		if change == nil {
+			continue
+		}
+
+		result := make(map[string]interface{})
+		result["block"] = change.Hash.String()
+		changes := [][]string{}
+		for _, v := range change.Changes {
+			kv := []string{common.BytesToHex(v.Key), common.BytesToHex(v.Value)}
+			changes = append(changes, kv)
+		}
+		result["changes"] = changes
+
+		res := newSubcriptionBaseResponseJSON()
+		res.Method = "state_storage"
+		res.Params.Result = result
+		res.Params.SubscriptionID = l.subID
+		l.wsconn.safeSend(res)
+	}
+}
+
+// BlockListener to handle listening for blocks importedChan
+type BlockListener struct {
+	channel chan *types.Block
+	wsconn  *WSConn
+	chanID  byte
+	subID   int
+}
+
+func (c *WSConn) initBlockListener(reqID float64) (int, error) {
+	bl := &BlockListener{
+		channel: make(chan *types.Block),
+		wsconn:  c,
+	}
+
+	if c.blockAPI == nil {
+		c.safeSendError(reqID, nil, "error BlockAPI not set")
+		return 0, fmt.Errorf("error BlockAPI not set")
+	}
+	chanID, err := c.blockAPI.RegisterImportedChannel(bl.channel)
+	if err != nil {
+		return 0, err
+	}
+	bl.chanID = chanID
+	c.qtyListeners++
+	bl.subID = c.qtyListeners
+	c.subscriptions[bl.subID] = bl
+	c.blockSubChannels[bl.subID] = chanID
+	initRes := newSubscriptionResponseJSON(bl.subID, reqID)
+	c.safeSend(initRes)
+
+	return bl.subID, nil
+}
+
+// Listen implementation of Listen interface to listen for importedChan changes
+func (l *BlockListener) Listen() {
+	for block := range l.channel {
+		if block == nil {
+			continue
+		}
+		head, err := modules.HeaderToJSON(*block.Header)
+		if err != nil {
+			logger.Error("failed to convert header to JSON", "error", err)
+		}
+
+		res := newSubcriptionBaseResponseJSON()
+		res.Method = "chain_newHead"
+		res.Params.Result = head
+		res.Params.SubscriptionID = l.subID
+		l.wsconn.safeSend(res)
+	}
+}
+
+// BlockFinalizedListener to handle listening for finalized blocks
+type BlockFinalizedListener struct {
+	channel chan *types.Header
+	wsconn  *WSConn
+	chanID  byte
+	subID   int
+}
+
+func (c *WSConn) initBlockFinalizedListener(reqID float64) (int, error) {
+	bfl := &BlockFinalizedListener{
+		channel: make(chan *types.Header),
```

Comment:
Why are all subscription channels unbuffered?

Reply:
No reason, I can add a buffer,  any suggested size?  256?

Reply:
Yes. I think 256 would work fine. We need to make all the listener channels buffered. 
Unbuffered channels are usually slow due to their blocking nature. Here is a [benchmark report](https://tpaschalis.github.io/channels-limitations-speed/) for buffered vs unbuffered channels.
```
BenchmarkBufferedChannelEmptyStruct-8         23657084            49.9 ns/op
BenchmarkBufferedChannelOneByte-8             21230530            54.6 ns/op      18.31 MB/s

BenchmarkUnbufferedChannelEmptyStruct-8        6075384            177 ns/op
BenchmarkUnbufferedChannelOneByte-8            6341457            184 ns/op       5.44 MB/s
```
I have created a separate issue for this. https://github.com/ChainSafe/gossamer/issues/1478